File: ibkr_show_frontend/src/types/dailyPositionReview.ts
import type { AgentRunTraceItem } from './tradeDecision'

export interface DailyPositionReviewHealth {
  enabled: boolean
  llm_configured: boolean
  longbridge_configured: boolean
  account_data_source: string
  public_market_data_source: string
  message: string
}

export interface DailyPositionReviewPositionItem {
  symbol: string
  normalized_symbol: string
  name: string | null
  asset_class: string | null
  sub_category: string | null
  quantity: number | null
  mark_price: number | null
  market_value: number | null
  weight: number | null
  daily_change_percent: number | null
  daily_pnl: number | null
  contribution_ratio: number | null
  average_cost: number | null
  cost_basis: number | null
  unrealized_pnl: number | null
  unrealized_pnl_percent: number | null
  is_major_contributor: boolean
  is_major_drag: boolean
  data_source: string
}

export interface DailyPositionReviewContext {
  report_date: string
  data_sources: Record<string, string>
  overview: {
    report_date: string
    currency?: string | null
    total_equity?: number | null
    daily_pnl?: number | null
    daily_return_percent?: number | null
    total_position_value?: number | null
    cash?: number | null
    cash_ratio?: number | null
    position_count: number
    top_contributors: DailyPositionReviewPositionItem[]
    top_drags: DailyPositionReviewPositionItem[]
    summary: string
    ibkr_pnl_breakdown: Record<string, number | null>
  }
  positions: DailyPositionReviewPositionItem[]
  rankings: Record<string, DailyPositionReviewPositionItem[]>
  risk: {
    max_position?: DailyPositionReviewPositionItem | null
    max_single_position_weight?: number | null
    top3_weight?: number | null
    top5_weight?: number | null
    theme_buckets: Array<Record<string, unknown>>
    semiconductor_ai_tech_weight?: number | null
    cash_ratio?: number | null
    max_position_down_5pct_account_impact_percent?: number | null
    risk_flags: string[]
    account_posture?: string | null
  }
  benchmarks: {
    items: Array<{
      symbol: string
      return_percent: number | null
      account_excess_return_percent: number | null
      source: string
    }>
    beta_alpha_note: string
  }
  focus_symbols: string[]
  attribution_quality: Record<string, unknown>
  data_quality: {
    missing_fields?: string[]
    warnings?: string[]
  }
}

export interface DailyPositionReviewResult {
  id: string
  report_date: string
  review_type: string
  summary: string
  account_conclusion: string
  attribution_summary: string
  major_contributors_analysis: Array<Record<string, unknown>>
  major_drags_analysis: Array<Record<string, unknown>>
  focus_symbol_analyses: Array<Record<string, unknown>>
  market_context: string
  risk_analysis: string
  tomorrow_watchlist: Array<Record<string, unknown>>
  operation_observation: string
  data_limitations: string[]
  evidence_used: string[]
  data_source_summary: Record<string, string>
  deterministic_context: Record<string, unknown>
  display_context: Record<string, unknown>
  run_trace: AgentRunTraceItem[]
  metadata: Record<string, unknown>
  evidence_summary: Record<string, unknown>
  run_trace_summary: Record<string, unknown>
  created_at: string
  updated_at: string
}

export interface DailyPositionReviewDateListResponse {
  items: string[]
}

export interface DailyPositionReviewListResponse {
  items: DailyPositionReviewResult[]
}

File: ibkr_show_backend/app/schemas/daily_position_review.py
from pydantic import BaseModel, Field

from app.schemas.trade_decision import AgentRunTraceItem


class DailyPositionReviewHealthResponse(BaseModel):
    enabled: bool
    llm_configured: bool
    longbridge_configured: bool
    account_data_source: str
    public_market_data_source: str
    message: str
    agent_mode: str = "daily_position_review_langgraph_v1"
    graph_version: str = "daily_position_review_graph_v1"


class DailyPositionReviewGenerateRequest(BaseModel):
    force_refresh: bool = False


class DailyPositionReviewDateListResponse(BaseModel):
    items: list[str]


class DailyPositionReviewPositionItem(BaseModel):
    symbol: str
    normalized_symbol: str
    name: str | None = None
    asset_class: str | None = None
    sub_category: str | None = None
    quantity: float | None = None
    mark_price: float | None = None
    market_value: float | None = None
    weight: float | None = None
    daily_change_percent: float | None = None
    daily_pnl: float | None = None
    contribution_ratio: float | None = None
    average_cost: float | None = None
    cost_basis: float | None = None
    unrealized_pnl: float | None = None
    unrealized_pnl_percent: float | None = None
    is_major_contributor: bool = False
    is_major_drag: bool = False
    data_source: str = "IBKR"


class DailyPositionReviewOverviewResponse(BaseModel):
    report_date: str
    currency: str | None = None
    total_equity: float | None = None
    daily_pnl: float | None = None
    daily_pnl_source: str | None = None
    daily_return_percent: float | None = None
    total_position_value: float | None = None
    cash: float | None = None
    cash_ratio: float | None = None
    position_count: int = 0
    top_contributors: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_drags: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    summary: str
    ibkr_pnl_breakdown: dict = Field(default_factory=dict)


class DailyPositionReviewPositionsResponse(BaseModel):
    report_date: str
    items: list[DailyPositionReviewPositionItem]


class DailyPositionReviewRankingsResponse(BaseModel):
    report_date: str
    profit_contributors: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    loss_drags: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_gainers: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_losers: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_weights: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_unrealized_gains: list[DailyPositionReviewPositionItem] = Field(default_factory=list)
    top_unrealized_losses: list[DailyPositionReviewPositionItem] = Field(default_factory=list)


class DailyPositionReviewRiskResponse(BaseModel):
    report_date: str
    max_position: dict | None = None
    max_single_position_weight: float | None = None
    top3_weight: float | None = None
    top5_weight: float | None = None
    theme_buckets: list[dict] = Field(default_factory=list)
    semiconductor_ai_tech_weight: float | None = None
    cash_ratio: float | None = None
    max_position_down_5pct_account_impact_percent: float | None = None
    risk_flags: list[str] = Field(default_factory=list)
    account_posture: str | None = None


class DailyPositionReviewContextResponse(BaseModel):
    report_date: str
    data_sources: dict
    overview: dict
    positions: list[DailyPositionReviewPositionItem]
    rankings: dict
    risk: dict
    benchmarks: dict
    focus_symbols: list[str]
    attribution_quality: dict
    data_quality: dict


class DailyPositionReviewResult(BaseModel):
    id: str
    report_date: str
    review_type: str
    summary: str
    account_conclusion: str
    attribution_summary: str
    major_contributors_analysis: list[dict]
    major_drags_analysis: list[dict]
    focus_symbol_analyses: list[dict]
    market_context: str
    risk_analysis: str
    tomorrow_watchlist: list[dict]
    operation_observation: str
    data_limitations: list[str]
    evidence_used: list[str]
    data_source_summary: dict
    deterministic_context: dict = Field(default_factory=dict)
    run_trace: list[AgentRunTraceItem] = Field(default_factory=list)
    metadata: dict = Field(default_factory=dict)
    evidence_summary: dict = Field(default_factory=dict)
    run_trace_summary: dict = Field(default_factory=dict)
    subagent_card_pack: dict = Field(default_factory=dict)
    subagent_trace: dict = Field(default_factory=dict)
    evidence_card_summary: dict = Field(default_factory=dict)
    graph_node_traces: list[dict] = Field(default_factory=list)
    graph_version: str | None = None
    fallback_used: bool = False
    fallback_reason: str | None = None
    status: str | None = None
    created_at: str
    updated_at: str


class DailyPositionReviewListResponse(BaseModel):
    items: list[DailyPositionReviewResult]
